### ISSUE TITLE
Crash on commit with deleted documents

### ISSUE THREAD

**Issue by brainlock** [quickwit-oss/tantivy#681]
The following sequence of events:

    - start with an empty index
    - add and then delete a document
    - add at least 32 documents
    - commit
    - commit again

leads to a crash on the second commit operation

```
thread 'segment_updater0' panicked at 'index out of bounds: the len is 4 but the index is 4', src/fastfield/delete.rs:65:21
```

Repro code:
```rust
        let mut schema_builder = schema::Schema::builder();
        let text_field = schema_builder.add_text_field("text", schema::TEXT);
        let index = Index::create_from_tempdir(schema_builder.build()).unwrap();

        let mut index_writer = index.writer(3_000_000).unwrap();

        // there must be one deleted document in the segment
        index_writer.add_document(doc!(text_field=>"b"));
        index_writer.delete_term(Term::from_field_text(text_field, "b"));

        // we need enough data to trigger the bug (at least 32 documents)
        for _ in 0..32 {
            index_writer.add_document(doc!(text_field=>"c"));
        }

        // a delete here would fix up the internal data structure that is being
        // accessed out of bounds (would make it big enough)

        // first commit: fine
        index_writer.commit().unwrap();

        // second commit: crash
        // in the first commit, there was no "previous deleted bitset" to check
        index_writer.commit().unwrap();
```

PR with a fix incoming.

**Comment by fulmicoton** [comment]
@brainlock Awesome bug report thanks! I was able to reproduce in no time thanks to your test!

**Comment by brainlock** [comment]
After the PR with the simple fix, I took a stab at removing the iteration from 0 .. max_doc. This was going to be another one of my "version A or version B" questions, except this time I don't completely like either of them:

- with the iterator API: https://github.com/brainlock/tantivy/commit/8213fce1d2632b0c58f2bd7ef77718f0e666c06b the code is very simple and convincing, the part I don't like is that `Vec::new` in the inner iterator

- with an explicit iterator to hold all the state at once: https://github.com/brainlock/tantivy/commit/9f82861e25cf40135648cefd39afdaa65d1983dd  but this one is so much more code that I'm not sure it's worth it, unless proven really faster than 0 .. max_doc for common workloads...

what I would have liked (but I'm hitting the limits of my Rust here) is version A, but with a way to keep the state of the inner iterator and not allocating a new vec each time.

Anyway this is not strictly related to the bug so feel free to ignore :)

**Comment by fulmicoton** [comment]
@brainlock  I investigated the root cause and addressed. Can you review the fix in  #683 ?

**Comment by brainlock** [comment]
> @brainlock I investigated the root cause and addressed. Can you review the fix in #683 ?

@fulmicoton late review, but LGTM :) Thanks!
